### DISCUSSION TITLE
X11 app with Libadwaita slow as hell

### DISCUSSION THREAD

**Discussion by funsafe-ptr** [termux/termux-packages#22494] [Q&A] [answered]
Do you know why? You can try installing an app that uses libadwaita, or try install libadwaita-demos. They are noticeably heavy, don’t respect the theme, have a huge UI, and some big ugly black borders. These borders seem to be fixed if the compositor is running, I'm not running desktop environment though, just openbox as my window manager.

![ss](https://github.com/user-attachments/assets/68e0bd56-cce2-4d55-a75e-bfbc1b68bf24)

**Comment by licy183** [comment]
IIRC a similar issue is reported #21321. But sadly I don't know why...

**Comment by Biswa96** [comment]
I can reproduce the issue also. The adwaita demo feels slower compared to other GUI programs in xfce4 in termux-x11. Though, I did not see that black border like artifacts.

**Reply by funsafe-ptr** [reply]
A black border shows up if the compositor is disabled. I think libdwaita uses compositor transparency for shadowing. I tried the solution from https://github.com/i3/i3/discussions/5896, and setting the environment variable worked Although it doesn’t fit very well, the window doesn’t have a visible border. It's still slow af.
```fish
GTK_THEME=Matcha-dark-aliz adwaita-1-demo &
```
![Screenshot_2024-12-10-21-13-23-82](https://github.com/user-attachments/assets/f96d9234-88f4-41a4-8acf-51e5feecdabb)

**Comment by funsafe-ptr** [comment, marked as answer]
Turns out the slowness is because it's heavily using the GPU, and I'm currently using llvmpipe software renderer. Switching to virpipe helps a bit, but not much. For now, I'm just avoiding the app and find the alternative.

**Comment by Biswa96** [comment]
There is a merge request to add new Android backend in gtk https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/7555 (thanks Tragicomedy). I wonder if it can be added in termux, something like termux:api.
